src/filesystem.cpp

```diff
@@ -48,15 +48,147 @@ bool FileSystemAccess::islchex(char c) const
+        return FS_DEFAULT;
+    }
+
+#if defined (__linux__) || defined (__ANDROID__)
+    struct statfs fileStat;
+    if (!statfs(dstPath->c_str(), &fileStat))
+    {
+        switch (fileStat.f_type)
+        {
+            case EXT2_SUPER_MAGIC:
+                return FS_UNIX;
+            case MSDOS_SUPER_MAGIC:
+                return FS_FAT32;
+            case HFS_SUPER_MAGIC:
+                return FS_APPLE;
+            case NTFS_SB_MAGIC:
+                return FS_WIN;
+            default:
+                return FS_DEFAULT;
+        }
+    }
+#elif defined  (__APPLE__)
+    struct statfs fileStat;
+    if (!statfs(dstPath->c_str(), &fileStat))
+    {
+        if (!strcmp(fileStat.f_fstypename, "apfs")
+                || !strcmp(fileStat.f_fstypename, "hfs"))
+        {
+            return FS_APPLE;
+        }
+        if (!strcmp(fileStat.f_fstypename, "ntfs"))
+        {
+            return FS_WIN;
+        }
+        if (!strcmp(fileStat.f_fstypename, "msdos"))
+        {
+            return FS_FAT32;
+        }
+    }
+#elif defined(_WIN32) || defined(_WIN64)
+    TCHAR volumeName[MAX_PATH + 1] = { 0 };
+    TCHAR fileSystemName[MAX_PATH + 1] = { 0 };
+    DWORD serialNumber = 0;
+    DWORD maxComponentLen = 0;
+    DWORD fileSystemFlags = 0;
+
+    if (GetVolumeInformationA(dstPath->c_str(), volumeName, sizeof(volumeName),
+                             &serialNumber, &maxComponentLen, &fileSystemFlags,
+                             fileSystemName, sizeof(fileSystemName)) == true)
+    {
+        if (!strcmp(fileSystemName, "NTFS")
+                || !strcmp(fileSystemName, "exFAT"))
+        {
+            return FS_WIN;
+        }
+        else if (!strcmp(fileSystemName, "FAT32"))
+        {
+            return FS_FAT32;
+        }
+    }
+#endif
+    return FS_DEFAULT;
+}
+
+bool FileSystemAccess::isControlChar(unsigned char c) const
+{
+    return (c <= '\x1F' || (c >= '\x7F' && c <= '\x9F'));
+}
+
+// Group different filesystems types in families, according to it's restricted charsets
+bool FileSystemAccess::islocalfscompatible(unsigned char c, int fileSystemType) const
+{
+    if (isControlChar(c))
+    {
+        return false;
+    }
+
+    switch (fileSystemType)
+    {
+        case FS_APPLE:
+            // APFS, HFS, HFS+ restricted characters => :
+            return c != '\x3A';
+        case FS_UNIX:
+            // ext2/ext3/ext4 restricted characters =>  /
+            return c != '\x2F';
+        case FS_FAT32:
+            // FAT32 restricted characters => 0x0000-0x001F 0x007F " * / : < > ? \ | + , . ; = [ ]
+            return !strchr("\\/:?\"<>|*+,.;=[]", c);
+        case FS_WIN:
+        default:
+            // NTFS restricted characters => 0x0000-0x001F 0x007F " * / : < > ? \ |
+            // If filesystem couldn't be detected we'll use a restrictive charset to avoid issues.
+            return !strchr("\\/:?\"<>|*", c);
+    }
+}
+
+string FileSystemAccess::getValidPath(const string *path) const
+{
```

Comment:
Consider to improve performance of this method, since it will be called multiple times